src/Features/LensEffects.cpp

```diff
@@ -0,0 +1,1110 @@
+		useCloudLUT = false;
+	}
+
+	overrideShader = false;
+}
+
+void LensEffects::BypassShader()
+{
+	auto context = globals::d3d::context;
+
+	context->VSSetShader(BypassVertexShader, NULL, NULL);
+	context->PSSetShader(NULL, NULL, NULL);
+
+	overrideShader = false;
+}
+
+void LensEffects::GetWeatherShader()
+{
+	auto sky = globals::game::sky;
+
+	if (precipCurrent) {
+		RE::TESWeather* weather = (weatherCurrent) ? sky->currentWeather : sky->lastWeather;
+		if (weather->data.flags.any(RE::TESWeather::WeatherDataFlag::kSnow) && shaderdesc == Shaders::LensIce) {
+			auto factor = 0.0f;
+			if (weatherCurrent) {
+				factor = std::clamp(sky->currentWeatherPct - settings.coldsettings.LI_FadeIn, 0.0f, 1.0f);
+				currentPrecip.y = 0.0f;
+			} else {
+				factor = std::clamp(sky->currentWeatherPct - settings.coldsettings.LI_FadeOut, 0.0f, 1.0f);
+			}
+
+			if (factor < 0.01f || RE::PlayerCharacter::GetSingleton()->GetParentCell()->IsInteriorCell())
+				precipstart = std::chrono::high_resolution_clock::now();
+
+			else if (factor > 0.01f) {
+				auto time = std::chrono::high_resolution_clock::now();
+				float elapsed = std::chrono::duration_cast<std::chrono::duration<float>>(time - precipstart).count();
+				auto precipfade = std::clamp(elapsed / settings.mainsettings.LI_FadeDuration, 0.0f, 1.0f);
+				currentPrecip.y = (weatherCurrent) ? precipfade : 1.0f - precipfade;
+			}
+
+			if (RE::PlayerCharacter::GetSingleton()->GetParentCell()->IsInteriorCell() == false)
+				return;
+		}
+	}
+
+	shaderdesc = Shaders::Bypass;
+}
+
+void LensEffects::UpdateWeatherBasedDisable()
+{
+	static bool currWeatherRole = false;
+	static bool prevWeatherRole = false;
+
+	if (CheckWeatherChange()) {
+		disableSunFX = false;
+		weatherFadeout = 0.0f;
+		currWeatherRole = std::binary_search(weatherDisables.begin(), weatherDisables.end(), WeatherID);
+		prevWeatherRole = std::binary_search(weatherDisables.begin(), weatherDisables.end(), PrevWeatherID);
+	}
+
+	if (currWeatherRole || prevWeatherRole) {
+		auto weatherTransition = GetWeatherTransTime();
+		if ((weatherTransition == 1.0f && currWeatherRole) || (currWeatherRole && prevWeatherRole))
+			disableSunFX = true;
+		weatherFadeout = (currWeatherRole) ? weatherTransition : 1.0f - weatherTransition;
+	}
+}
+
+void LensEffects::UpdateFrameGenBasedDisable()
+{
+	auto method = globals::upscaling->GetUpscaleMethod();
+
+	if (globals::upscaling->IsFrameGenerationActive() || method == Upscaling::UpscaleMethod::kDLSS || method == Upscaling::UpscaleMethod::kFSR) {
+		renderdata->GetEffect(Shaders::LensIce).Toggle(false);
+		upscalingActive = true;
+	} else if (upscalingActive) {
+		renderdata->GetEffect(Shaders::LensIce).Toggle(settings.EnableIce);
+		upscalingActive = false;
+	}
+}
+
+DirectX::XMFLOAT4A LensEffects::GetSunPosition()
+{
+	auto sunWSPos = *skyrim_SunPosition;
+	float4 sunPosition = { sunWSPos.x, sunWSPos.y, sunWSPos.z, 1.0f };
+
+	Matrix viewMatrix = Util::GetCameraData(0).viewMat;
```

Comment:
While it may be ok, be aware of eye index which is 0 for flat or may need to be set for VR based on left or right eye. You don't have to setup for VR, but sometimes it just works if you adopt the existing methods.